### PR TITLE
Web application

### PR DESCRIPTION
**Pull request by certik**
I have finally created the GAE web app:

http://reviews.sympy.org/

and modified the `sympy-bot` to upload into this app instead of pastehtml.com. Source code of this app is in the `web` directory and it is separate from the build bot. It has to be uploaded to the google app engine.

It seems to be working quite well for me. Could you please test it by running the `sympy-bot` as usual (using my branch)? It will upload to reviews.sympy.org, and give you the URL at the end, so that you can see it.

For some example pull request, see: http://reviews.sympy.org/pullrequest/647/

### CODE REVIEW

**Comment by mattpap** [comment]
That's really cool.

A few suggestions:
1. On the main page I would add a title of every pull request, instead of just a number.
2. "Repository and branch: https://github.com/smichr/sympy polyx": I would make both repo and branch links (now only branch is a link).

**Comment by certik** [comment]
Thanks for the suggestions. I will implement those as an improvement after this is merged. The reason being that every change has to be thoroughly tested, so I want to make sure that the current state works robustly, push it in, and then improve.

**Comment by certik** [comment]
Another suggestion: sort the pull request by the "creation date" on the front page, and also split them to "can be merged", "can't be merged".

If you have more ideas, please share it. Eventually, we'll implement them, now when the basis is done, it should be quite easy.

**Comment by certik** [comment]
I think that the front page should emulate this page: https://github.com/sympy/sympy/pulls, just the part about pull requests. With all the info, just do it better and more usable. So there would be buttons like "mergeable", "open", "close", "recent", "sort by contributors 'newness' to sympy" and so on. And we can make it really usable and put a big link to sympy.org for it, so that people start reviewing things more.

**Comment by asmeurer** [comment]
It would be awesome if we could have all the important metadata in the webapp, and then even a search function there, which is something that is annoyingly missing from GitHub's interface. Indeed, if it were more or less like the GitHub page (though I could think of a few improvements :-), that would be great.

It would also be cool if it were full of html 5, though I realize that that's way more work :)

File: web/templates/pullrequest.html [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+<h2>Tests Summary</h2>
+Click on the individual test to get the full report with logs.
+<ul>
+{% for t in tasks %}
+    <li><a href="/report/{{ t.key }}">Test {{ forloop.counter }}</a>
+({% if t.result == "Passed" %}<span class="result_ok">Passed</span>,
+{% else %}<span class="result_fail">{{ t.result }}</span>,
+{% endif %}
+    command: <span class="code">{{ t.interpreter }} {{ t.testcommand }}</span>,
+    uploaded at: {{ t.uploaded_at }})</li>
```
Comment (asmeurer) [inline web/templates/pullrequest.html:33]: Could we also add "by whomever" (i.e., the person who did the review)?

**Comment by asmeurer** [comment]
I have seen webapps like http://githubissues.heroku.com/#sympy/sympy/ for GitHub issues (GitHub considers pull requests to be issues).  We could potentially take one of those and modify it if it's open source.

**Comment by mattpap** [comment]
> I have seen webapps like http://githubissues.heroku.com/#sympy/sympy/ for GitHub issues

This particular one is more annoying than useful (it's extremely slow).

**Comment by asmeurer** [comment]
Can you add to the readme that this requires simplejson?

**Comment by mattpap** [comment]
> Can you add to the readme that this requires simplejson?

New Python interpreters have `json` module built-in, which is equivalent to `simplejson`. Just use:

``` python
try:
    import json
except ImportError:
    import simplejson as json
```

**Comment by certik** [comment]
I found another bug:

http://reviews.sympy.org/report/agZzeW1weTNyCwsSBFRhc2sYiHUM

if I upload a test result, before the app updates the list of pull requests from github, the pull request number is missing. Internally, if I can't find the pull request in the database, I should create the pull request, with the given number anyway (in the database). The update will then just update the data about it, so all will work.

File: web/templates/pullrequest.html (hunk 1 above)
Comment (certik) [inline web/templates/pullrequest.html:33]: Yes, also together with the picture and put everything into a nice table. And add buttons to sort by various criteria.

**Comment by certik** [comment]
I thought I am using the import trick with json. Anyway, this is easy to fix.

Can you both try to use it to upload a test result? Just to make sure, that things work. If so, I'll merge this and create issues for the things you raised.

**Comment by asmeurer** [comment]
The page http://reviews.sympy.org/report/ gives a traceback, and that report you linked to has the `#` where the pull request number should be as a link to http://reviews.sympy.org/pullrequest/, which doesn't exist (btw, this needs a 404 page).

**Comment by asmeurer** [comment]
Did you mean to push up a change?  It's still the same as before.

**Comment by mattpap** [comment]
btw. #648 (float_new) was merged but it's still on the front page (reviews.sympy.org).

**Comment by certik** [comment]
@asmeurer, I don't understand. 'http://reviews.sympy.org/report' gives a traceback, because it is an invalid url. You have to use url like this: 'http://reviews.sympy.org/report/agZzeW1weTNyCwsSBFRhc2sYhH0M'. Obviously, the error page should be better (nice 404, not a traceback), thanks for the tip, I'll put it into issues. I can see your uploads here:

http://reviews.sympy.org/pullrequest/648

so it seems to work well. 

The pull requests currently remain on the front page forever. I'll implement hiding the closed pull requests. Thanks for noticing.

**Comment by certik** [comment]
Anyway, I am merging it now, as I want to move further with this, I think it is well tested. I have created an issue #41 from all your tips. Thanks!

**Comment by mattpap** [comment]
btw. We have to implement automatic updates soon, because I submitted a review to pull request https://github.com/sympy/sympy/pull/652, but no review was uploaded here http://reviews.sympy.org/pullrequest/652/. I updated the main page after pull request review.

**Pull request merged**